src/ec2/model.ts

```diff
@@ -41,53 +41,55 @@ export class Ec2ConnectionManager {
     }
 
     protected async getAttachedPolicies(instanceId: string): Promise<IAM.attachedPoliciesListType> {
-        try {
-            const IamRole = await this.ec2Client.getAttachedIamRole(instanceId)
-            const iamResponse = await this.iamClient.listAttachedRolePolicies(IamRole!.Arn!)
-            return iamResponse.AttachedPolicies!
-        } catch (err) {
+        const IamRole = await this.ec2Client.getAttachedIamRole(instanceId)
+        if (!IamRole) {
             return []
         }
+        const iamResponse = await this.iamClient.listAttachedRolePolicies(IamRole!.Arn!)
+
+        return iamResponse.AttachedPolicies ?? []
     }
 
     public async hasProperPolicies(instanceId: string): Promise<boolean> {
         const attachedPolicies = (await this.getAttachedPolicies(instanceId)).map(policy => policy.PolicyName!)
         const requiredPolicies = ['AmazonSSMManagedInstanceCore', 'AmazonSSMManagedEC2InstanceDefaultPolicy']
 
-        return requiredPolicies.every(policy => attachedPolicies.includes(policy))
+        return requiredPolicies.length !== 0 && requiredPolicies.every(policy => attachedPolicies.includes(policy))
     }
 
-    public async handleStartSessionError(err: AWSError, selection: Ec2Selection): Promise<Error> {
-        const isInstanceRunning = (await this.ec2Client.getInstanceStatus(selection.instanceId)) == 'running'
+    private async isInstanceRunning(instanceId: string): Promise<boolean> {
+        return (await this.ec2Client.getInstanceStatus(instanceId)) == 'running'
+    }
+
+    private throwConnectionError(
+        message: string,
+        code: Ec2ConnectErrorCode,
+        selection: Ec2Selection,
+        documentationUri?: vscode.Uri
+    ) {
         const generalErrorMessage = `Unable to connect to target instance ${selection.instanceId} on region ${selection.region}. `
+        throw new ToolkitError(
+            generalErrorMessage + message,
+            documentationUri ? { code: code, documentationUri: documentationUri } : { code: code }
+        )
+    }
+
+    public async checkForStartSessionError(selection: Ec2Selection): Promise<void> {
```

Comment:
Pre-checking works although a reactive approach may be more robust. Let's leave this as-is and come back to it later if we have time.

Reply:
I originally had this as a reactive approach, but it had some edge cases. There were ways we could get the `ssm.startSession` to not throw any errors, but the terminal immediately crashes. Thats why I moved it to checking the errors before. Should I be looking for errors both before and after if this is the case?